store/sqlstore/upgrade.go

```diff
@@ -374,6 +374,10 @@ func UpgradeDatabaseToVersion48(sqlStore SqlStore) {
 }
 
 func UpgradeDatabaseToVersion49(sqlStore SqlStore) {
+	// This version of Mattermost includes an App-Layer migration which migrates from hard-coded roles configured by
+	// a number of parameters in `config.json` to a `Roles` table in the database. The migration code can be seen
+	// in the file `app/app.go` in the function `DoAdvancedPermissionsMigration()`.
```

Comment:
What's the context for not just doing this here, in the upgrade path?

Reply:
Short answer: It's necessary to do it in the app layer to avoid circular dependencies.

Reply:
See discussion on the original PR that landed this here for more context: https://github.com/mattermost/mattermost-server/pull/8159/files#diff-23903225b64d2e452b1d25db40371ff3R81